### PR TITLE
Use Faabric distributed coordination and merge operations

### PR DESCRIPTION
**Pull request by Shillaker**
Update Faasm multi-threaded code to use latest Faabric features, including distributed coordination (https://github.com/faasm/faabric/pull/161), snapshot merge operations (https://github.com/faasm/faabric/pull/142), and scheduler hints (https://github.com/faasm/faabric/pull/169).

Changes:

- Use snapshot merge operations in OpenMP, both for general shared variables, reduce results.
- Use Faabric `PointToPointGroup`s for OpenMP locking and barriers.
- Cache the `SchedulingDecision` for each level of an OpenMP application, then pass this as a hint when scheduling the same number of threads at the same depth. 
- Remove coordination logic from OpenMP `Level` class (now handled by `PointToPointGroup`).
- Replace `MutexManager` with Faabric `PointToPointGroup` in pthread code.

This is an update of https://github.com/faasm/faasm/pull/480 which had gone stale.

I also tidied up a couple of things:

- Remove `faabric` container from Faasm `docker-compose.yml`. We should use the `faabric` CLI container in the Faabric project itself.
- Remove unhelpful/ stale OpenMP doc.

### CODE REVIEW

File: include/conf/FaasmConfig.h [hunk 1]
```diff
@@ -2,6 +2,22 @@
...
+// hard-coded in the build. This variable is just here for reference and must be
+// updated to match the value in the build.
+#define STACK_SIZE (4 * ONE_MB_BYTES)
+#define THREAD_STACK_SIZE (2 * ONE_MB_BYTES)
+
+// Properties of dynamic modules. Heap size must be wasm-module-page-aligned.
+// One page is 64kB
+#define DYNAMIC_MODULE_STACK_SIZE (2 * ONE_MB_BYTES)
+#define DYNAMIC_MODULE_MEMORY_SIZE (66 * WASM_BYTES_PER_PAGE)
+#define GUARD_REGION_SIZE (10 * WASM_BYTES_PER_PAGE)
```
Comment (Shillaker) [inline include/conf/FaasmConfig.h:19]: Just moved this stuff out of the WAVM module header as it's shared configuration and useful elsewhere.

File: tests/test/wasm/test_snapshots.cpp [hunk 2]
```diff
@@ -337,106 +338,4 @@ TEST_CASE_METHOD(WasmSnapTestFixture,
 
     ::munmap(snapMemory, snapSize);
 }
-
-TEST_CASE_METHOD(WasmSnapTestFixture,
-                 "Test ignoring stacks",
```
Comment (Shillaker) [inline tests/test/wasm/test_snapshots.cpp:342]: This no longer happens.

File: README.md [hunk 3]
```diff
@@ -73,7 +73,7 @@ More detail on some key features and implementations can be found below:
 - [Kubernetes and Knative integration](docs/kubernetes.md)- deploying Faasm as part of a full serverless platform.
 - [Bare metal/ VM deployment](docs/bare_metal.md) - deploying Faasm on bare metal or VMs as a stand-alone system.
 - [API](docs/api.md) - invoking and managing functions and state through Faasm's HTTP API.
-- [MPI](docs/mpi.md) and [OpenMP](docs/openmp.md) - executing existing MPI and OpenMP applications in Faasm.
```
Comment (csegarragonz) [inline README.md:76]: Is there value in linking the OpenMp docs from the cpp repo here? Maybe worth for the future to have accessible documentation about MPI and OpenMP in Faasm.

File: tests/dist/threads/test_pthreads.cpp [hunk 4]
```diff
@@ -5,6 +5,7 @@
 #include <faabric/scheduler/Scheduler.h>
 
 namespace tests {
+
```
Comment (csegarragonz) [inline tests/dist/threads/test_pthreads.cpp:8]: I usually don't add this lines myself, does `clang-format` say something about it?

**Review by csegarragonz — APPROVED** [review]
:+1: Just two minor things.

File: tests/dist/threads/test_pthreads.cpp (hunk 4 above)
Comment (Shillaker) [reply tests/dist/threads/test_pthreads.cpp:8]: I don't see what harm this newline does, in fact, I think it makes it easier to read as it clearly separates the namespace declaration from whatever comes next. The Google style guide seems to agree as all their examples have a newline: https://google.github.io/styleguide/cppguide.html#Namespaces. 

However, doing a brief scan I can see that the repo is pretty mixed at the moment and can't find a rule in clang-tidy to enforce anything so we can leave it for now.

File: README.md (hunk 3 above)
Comment (Shillaker) [reply README.md:76]: If the OpenMP docs said anything about the Faasm implementation of OpenMP then there would be, however, those docs are purely about debugging issues with compilation, which is only relevant to the `cpp` repo.

**Pull request merged**